lib/Tools/circt-lec/ConstructLEC.cpp

```diff
@@ -0,0 +1,140 @@
+  auto global = moduleOp.lookupSymbol<LLVM::GlobalOp>(str);
+  if (!global) {
+    OpBuilder b = OpBuilder::atBlockEnd(moduleOp.getBody());
+    auto arrayTy = LLVM::LLVMArrayType::get(b.getI8Type(), str.size() + 1);
+    global = b.create<LLVM::GlobalOp>(
+        loc, arrayTy, /*isConstant=*/true, LLVM::linkage::Linkage::Private, str,
+        StringAttr::get(b.getContext(), Twine(str).concat(Twine('\00'))));
+  }
+
+  // FIXME: sanity check the fetched global: do all the attributes match what
+  // we expect?
+
+  return builder.create<LLVM::AddressOfOp>(loc, global);
+}
+
+hw::HWModuleOp ConstructLECPass::lookupModule(StringRef name) {
+  Operation *expectedModule = SymbolTable::lookupNearestSymbolFrom(
+      getOperation(), StringAttr::get(&getContext(), name));
+  if (!expectedModule || !isa<hw::HWModuleOp>(expectedModule)) {
+    getOperation().emitError("module named '") << name << "' not found";
+    return {};
+  }
+  return cast<hw::HWModuleOp>(expectedModule);
+}
+
+void ConstructLECPass::runOnOperation() {
+  // Create necessary function declarations and globals
+  OpBuilder builder = OpBuilder::atBlockEnd(getOperation().getBody());
+  Location loc = getOperation()->getLoc();
+  auto ptrTy = LLVM::LLVMPointerType::get(builder.getContext());
+  auto voidTy = LLVM::LLVMVoidType::get(&getContext());
+
+  // Lookup or declare printf function.
+  auto printfFunc =
+      LLVM::lookupOrCreateFn(getOperation(), "printf", ptrTy, voidTy, true);
+
+  // Lookup the modules.
+  auto moduleA = lookupModule(firstModule);
+  if (!moduleA)
+    return signalPassFailure();
+  auto moduleB = lookupModule(secondModule);
+  if (!moduleB)
+    return signalPassFailure();
+
+  if (moduleA.getModuleType() != moduleB.getModuleType()) {
+    moduleA.emitError("module's IO types don't match second module's: ")
+        << moduleA.getModuleType() << " vs " << moduleB.getModuleType();
+    return signalPassFailure();
+  }
+
+  // Reuse the name of the first module for the entry function, so we don't have
+  // to do any uniquing and the LEC driver also already knows this name.
+  FunctionType functionType = FunctionType::get(&getContext(), {}, {});
+  func::FuncOp entryFunc =
+      builder.create<func::FuncOp>(loc, firstModule, functionType);
+
+  if (insertMainFunc) {
+    OpBuilder::InsertionGuard guard(builder);
+    auto i32Ty = builder.getI32Type();
+    auto mainFunc = builder.create<func::FuncOp>(
+        loc, "main", builder.getFunctionType({i32Ty, ptrTy}, {i32Ty}));
+    builder.createBlock(&mainFunc.getBody(), {}, {i32Ty, ptrTy}, {loc, loc});
+    builder.create<func::CallOp>(loc, entryFunc, ValueRange{});
+    // TODO: don't use LLVM here
+    Value constZero = builder.create<LLVM::ConstantOp>(loc, i32Ty, 0);
+    builder.create<func::ReturnOp>(loc, constZero);
+  }
+
+  builder.createBlock(&entryFunc.getBody());
+  auto lecOp = builder.create<verif::LogicEquivalenceCheckingOp>(loc);
+  auto *yieldA = moduleA.getBodyBlock()->getTerminator();
```

Comment:
Nit: Happy to be convinced otherwise but I think yieldA could be a slightly confusing name here, when you're creating yield ops to yield A and B further down - maybe something that makes it clearer that they're the original output ops?